master/internal/core_checkpoint_upstream_intg_test.go

```diff
@@ -0,0 +1,241 @@
+	region, err := getS3BucketRegion(context.TODO(), bucket)
+	if err != nil {
+		return err
+	}
+	sess, err := session.NewSession(&aws.Config{
+		Region: &region,
+	})
+	if err != nil {
+		return err
+	}
+	s3client := s3.New(sess)
+
+	for k, v := range mockCheckpointContent {
+		_, err = s3client.PutObject(&s3.PutObjectInput{
+			Bucket: aws.String(bucket),
+			Key:    aws.String(prefix + "/" + k),
+			Body:   aws.ReadSeekCloser(strings.NewReader(v)),
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func checkTgz(t *testing.T, content io.Reader, id string) {
+	zr, err := gzip.NewReader(content)
+	require.NoError(t, err, "failed to create a gzip reader")
+	tr := tar.NewReader(zr)
+	gotMap := make(map[string]string)
+	prefix := S3_TEST_PREFIX + "/" + id + "/"
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break // End of archive
+		}
+		require.NoError(t, err, "failed to read record header")
+		pathname := strings.TrimPrefix(hdr.Name, prefix)
+		buf := &strings.Builder{}
+		if hdr.Size > 0 {
+			_, err := io.Copy(buf, tr)
+			_, err = io.Copy(os.Stdout, tr)
+			require.NoError(t, err, "failed to read content of file", hdr.Name)
+		}
+		gotMap[pathname] = buf.String()
+	}
+	require.Equal(t, mockCheckpointContent, gotMap)
+}
+
+func checkZip(t *testing.T, content string, id string) {
+	zr, err := zip.NewReader(strings.NewReader(content), int64(len(content)))
+	require.NoError(t, err, "failed to create a zip reader")
+	gotMap := make(map[string]string)
+	prefix := S3_TEST_PREFIX + "/" + id + "/"
+	for _, f := range zr.File {
+		pathname := strings.TrimPrefix(f.Name, prefix)
+		buf := &strings.Builder{}
+		rc, err := f.Open()
+		require.NoError(t, err, "unable to decompress file", f.Name)
+		_, err = io.Copy(buf, rc)
+		require.NoError(t, err, "unable to read content of file", f.Name)
+		rc.Close()
+		gotMap[pathname] = buf.String()
+	}
+	require.Equal(t, mockCheckpointContent, gotMap)
+}
+
+func addMockCheckpointDB(t *testing.T, pgDB *db.PgDB, id uuid.UUID) {
+	etc.SetRootPath(db.RootFromDB)
+	user := db.RequireMockUser(t, pgDB)
+	// Using a different path than DefaultTestSrcPath since we are one level up than most db tests
+	exp := mockExperimentS3(t, pgDB, user, "../../examples/tutorials/mnist_pytorch")
+	tr := db.RequireMockTrial(t, pgDB, exp)
+	allocation := db.RequireMockAllocation(t, pgDB, tr.TaskID)
+	// Create checkpoints
+	checkpoint := db.MockModelCheckpoint(id, tr, allocation)
+	err := pgDB.AddCheckpointMetadata(context.TODO(), &checkpoint)
+	require.NoError(t, err)
+}
+
+func createCheckpoint(t *testing.T, pgDB *db.PgDB) (string, error) {
+	id := uuid.New()
+	addMockCheckpointDB(t, pgDB, id)
+	err := createMockCheckpointS3(S3_TEST_BUCKET, S3_TEST_PREFIX+"/"+id.String())
+	return id.String(), err
+}
+
+func newEchoContext() (echo.Context, *httptest.ResponseRecorder) {
+	e := echo.New()
+	rec := httptest.NewRecorder()
+	return e.NewContext(nil, rec), rec
+}
+
+func TestGetCheckpointEcho(t *testing.T) {
```

Comment:
Is it possible to not run the whole echo server here?  That makes this test harder to parse.

Would it be easier to just call BuildWriterPipeline and NewDownloader directly?

Other than that, this test is really detailed and has great checks with it.  I like it a lot.

Reply:
I kind of prefer running the whole server. I am less experienced with streaming and would like to test it e2e. The code outside buidWriterPipeline and NewDownloader is also pretty important. E.g., the downloaded file would be truncated if Close() is not called. Are there other ways I can make the test easier to parse?

Reply:
if you like it how it is, I don't feel strongly.